bugsnag-android-core/src/main/java/com/bugsnag/android/internal/DocumentPath.kt

```diff
@@ -0,0 +1,254 @@
+            require(value is MutableMap<*, *>) { "Value replacing document must be a mutable map" }
+            @Suppress("UNCHECKED_CAST")
+            return value as MutableMap<String, Any>
+        }
+        val filledInDocument = fillInMissingContainers(document, 0)
+        require(filledInDocument is MutableMap<*, *>) { "Document path must result in a top level map" }
+        @Suppress("UNCHECKED_CAST")
+        val updatedDocument = filledInDocument as MutableMap<String, Any>
+
+        applyValueToDocument(value, updatedDocument)
+        return updatedDocument
+    }
+
+    /**
+     * Fill in the document parts that will be navigated, but don't exist yet.
+     *
+     * For example, the path "foo.bar.1.-1" requires a document consisting of:
+     * - Map with key "foo" containing:
+     * - Map with key "bar" containing:
+     * - List with index 1 present, containing:
+     * - List where the last index is what we want to modify.
+     *
+     * Before we can modify the last index, we need to create all of the parent
+     * containers (if they don't exist yet). This function recursively descends
+     * the document, creating all necessary containers to make this path navigatable.
+     *
+     * @param parent The parent container (which might be null if it doesn't exist yet)
+     * @param index The index of the directive to use if a new container is necessary.
+     * @return The container that should be in place of the parent passed in (a list or a map).
+     */
+    private fun fillInMissingContainers(parent: Any?, index: Int): Any {
+        val directive = directives[index]
+        val updatedParent = when (parent) {
+            null -> directive.newContainer()
+            else -> parent
+        }
+
+        if (index + 1 < directives.size) {
+            val result =
+                fillInMissingContainers(directive.getFromContainer(updatedParent), index + 1)
+            directive.setInContainer(updatedParent, result)
+        }
+        return updatedParent
+    }
+
+    /**
+     * Navigate down into a document's sub-containers, then apply the specified value.
+     *
+     * @param value The value to apply (null means delete)
+     * @param document the document to apply this value to.
+     */
+    private fun applyValueToDocument(value: Any?, document: MutableMap<String, Any>) {
+        // Directives 0 through n-1 navigate to where we want to edit the document.
+        var currentContainer: Any = document
+        for (i in 0 until directives.size - 1) {
+            val directive = directives[i]
+            currentContainer = directive.getFromContainer(currentContainer)!!
+        }
+
+        // Only the last directive does the actual setting.
+        val directive = directives.last()
+        directive.setInContainer(currentContainer, value)
+    }
+
+    companion object {
+
+        const val escapeChar = '\\'
+        private const val pathSeparator = '.'
```

Comment:
```suggestion
        private const val ESCAPE_CHAR = '\\'
        private const val PATH_SEPARATOR = '.'
```

Kotlin's [naming convention for constants](https://kotlinlang.org/docs/coding-conventions.html#property-names) is similar to Java